start a graceful shutdown on connection recovery

mmmries opened the pull request:
An attempt to resolve #97 

I noticed this same problem in the form of a hanging travis build. My test suite includes an integration test where I force the connection to be closed via the management plugin and watch to make sure it reconnects and continues to receive messages from the broker.

Often times after my full suite ran the process would hang rather than exiting cleanly. This turned out to be because there was still a threadpool hanging around waiting for more work to do.

This fix initiates a graceful shutdown of the executor pool when a new connection is being established. There is no attempt to provide a timeout and a `shutDownNow()`, because as @michaelklishin mentioned in #97 we won't have any idea of how long we should wait around. Instead this just initiates a graceful shutdown and we hope for the best (ie the threads will complete or die before the process needs to exit). This is not an ideal implementation, but it resolves the common case for my usage where a connection reset might happen many hours before I actually shut down the process. That provides plenty of time for the threadpool to be cleaned up.

I would love to hear other ideas on a cleaner solution if anyone has them.

/cc @abrandoned @michaelklishin @ivoanjo

mmmries commented:
While checking the status of the travis builds I noticed that travis was using `1.7.19` and `9.0.0.0`. These are both fairly old at this point, so I updated to use `1.7.26` and `9.1.5.0`. I can remove that commit from this PR if you want, but it seemed like an easy thing to fix quickly.

michaelklishin commented on lib/march_hare/session.rb:
```diff
@@ -507,6 +507,7 @@ def new_connection_impl(hosts, host_selector)
 
       converting_rjc_exceptions_to_ruby do
         if @executor_factory
+          @executor.shutdown() if @executor
```
I suggest using [awaitTermination](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html#awaitTermination(long,%20java.util.concurrent.TimeUnit)) with a timeout (say, of 30 seconds by default).

mmmries commented:
@michaelklishin I added an `awaitTimeout` call and made it an option that can be passed in to `Session.connect`. It is specified in seconds to match the unit that is specified for heartbeats, network recovery interval etc, but I transform it to a number of milliseconds when doing the awaitTermination so user's can specify timeouts with partial seconds if they want.

I also added a `shutdownNow` if we timeout waiting for the pool to finish.Let me know what you think.

mmmries commented:
BTW I tested this with a consumer that does `sleep 10.0` for each message before trying to acknowledge it and it waits correctly before re-establishing the message. I end up getting error messages about trying to acknowledge a message on a channel that is already closed, but I suppose that's the expected behavior. If the connection is gone then I can't send back acknowledgments or rejections, but I'll rely on rabbit to re-send the message since it didn't get my acknowledgement.

```
#<MarchHare::ChannelAlreadyClosed: #method<connection.close>(reply-code=320, reply-text=CONNECTION_FORCED - Closed via management plugin, class-id=0, method-id=0)>
/Users/mmmries/code/march_hare/lib/march_hare/exceptions.rb:121:in `convert_and_reraise'
/Users/mmmries/code/march_hare/lib/march_hare/channel.rb:980:in `converting_rjc_exceptions_to_ruby'
/Users/mmmries/code/march_hare/lib/march_hare/channel.rb:739:in `basic_reject'
/Users/mmmries/code/march_hare/lib/march_hare/channel.rb:682:in `block in reject'
/Users/mmmries/code/march_hare/lib/march_hare/channel.rb:995:in `guarding_against_stale_delivery_tags'
/Users/mmmries/code/march_hare/lib/march_hare/channel.rb:681:in `reject'
/Users/mmmries/code/action_subscriber/lib/action_subscriber/middleware/env.rb:53:in `reject'
```

michaelklishin commented:
We have improved handling of acks/nacks in some clients when an ack happens on a *reopened* channel but clients can't do much about the case when a channel closed and not yet recovered. So yes, your understanding and expectations are solid and match mine :)

michaelklishin commented:
Thank you!
